### DISCUSSION TITLE
Long-press should cycle through checked, unchecked and missing

### DISCUSSION THREAD

**Discussion by lucasbru** [iSoron/uhabits#905] [Feature Requests] [answered]
If the "question-mark for missing"  feature is enabled, long-press on a habit/day cell should cycle through checked, unchecked and missing. Currently, there is no way to recover the "missing" state after you have (by mistake) long-pressed once on a habit.

Furthermore, you would need to integrate this somehow into the "edit checkmark history" feature inside the habit view.

**Comment by hiqua** [comment]
#902

**Comment by iSoron** [comment]
We don't currently allow cycling back to "unknown" to reduce the number of options the user has to cycle through. If anyone else thinks this is an important missing feature, feel free to upvote this feature request.

**Reply by ke352802081770314** [reply]
assuming that having to undo a "mis-click" is rare, having ? as the last state in the cycle would not make it worse for any other input.

**Comment by lucasbru** [comment]
Maybe cycling through all cases is not the best UX, maybe somebody has a
better idea. But having an irreversible action on a long-press is bad UX
for sure. There should be _some_ way to get back to unchecked state after
having accidentally long-tapped the wrong habit.

Alinson S. Xavier ***@***.***> schrieb am Mi., 19. Mai 2021,
03:53:

**Comment by Mrnofish** [comment]
After the switch to multiple states, "hold" to flip has become very cumbersome. As I'm interested in the extra states, I'm anticipating "hold" will be useless to me from now on.

When "tap" is enabled, it would be cool for long tap to offer some sort of direct state selection, perhaps through a pop-up menu as discussed previously. This would assuage concerns that there are too many states to go through before going back to "no data".

As "tap" also entails a higher risk of incorrectly changing the state of a "no data" day, it makes sense to me that that state should be available when short tap is enabled.

Since switching through states with "hold" takes quite some time, it also makes sense to me for it to be limited to just two states, so basically, the user would have a choice between "Advanced mode: simple tap and multi-state" or "Simple mode: hold and two states".

**Comment by lucasbru** [comment]
Yes, a little submenu for the different states would be really awesome.
Just like for selecting international keys on keyboards.

The submenu could always show up on long tap. The existing setting would
control whether short tap cycles between checked/unchecked states, or does
nothing.

>
>

**Comment by iSoron** [comment, marked as answer]
This has been implemented in the `dev` branch and should become available in Loop 2.1.

**Reply by lucasbru** [reply]
I tried it, it's perfect. Thanks

**Comment by DeWarner** [comment]
#1207 is a feature request for the popup menu idea mentioned by some comments in this thread

**Reply by Mrnofish** [reply]
Thanks Declan, upvoted.
